### PR TITLE
linuxPackages.gvusb2: init at 2020-07-27-unstable

### PR DESCRIPTION
**Pull request by djanatyn**
###### Motivation for this change

This is a new linux driver for the I-O DATA GV-USB2 capture card:
```
❯ find $(nix-build -A linuxPackages_latest.gvusb2) -type f
/nix/store/w77k64zb6v3x6sb9mwsf0n4qpjydkr6r-gvusb2-2020-07-27-unstable/lib/modules/5.10.7/extra/gvusb2-sound.ko
/nix/store/w77k64zb6v3x6sb9mwsf0n4qpjydkr6r-gvusb2-2020-07-27-unstable/lib/modules/5.10.7/extra/gvusb2-video.ko
❯ lsmod | grep gvusb2
gvusb2_video           32768  0
videobuf2_vmalloc      20480  1 gvusb2_video
videobuf2_v4l2         28672  1 gvusb2_video
videobuf2_common       53248  2 gvusb2_video,videobuf2_v4l2
videodev              253952  3 gvusb2_video,videobuf2_v4l2,videobuf2_common
gvusb2_sound           16384  0
snd_pcm               114688  6 snd_hda_codec_hdmi,snd_hda_intel,snd_hda_codec,snd_pcm_oss,gvusb2_sound,snd_hda_core
snd                    90112  17 snd_hda_codec_generic,snd_hda_codec_hdmi,snd_hwdep,snd_hda_intel,snd_hda_codec,snd_hda_codec_realtek,snd_timer,snd_pcm_oss,snd_pcm,gvusb2_sound,snd_mixer_oss
i2c_core               86016  7 videodev,drm_kms_helper,gvusb2_video,i2c_algo_bit,amdgpu,i2c_piix4,drm
usbcore               278528  6 xhci_hcd,gvusb2_video,usbhid,btusb,xhci_pci,gvusb2_sound
```

Using the driver, I was able to play Super Mario 64 on a CRT with simultaneous playback in VLC / OBS-Studio:
```
[  226.481672] usb 1-3: new high-speed USB device number 8 using xhci_hcd
[  226.683104] usb 1-3: config 1 interface 0 altsetting 0 endpoint 0x81 has invalid wMaxPacketSize 0
[  226.724103] usb 1-3: New USB device found, idVendor=04bb, idProduct=0532, bcdDevice= 0.05
[  226.724104] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=10
[  226.724105] usb 1-3: Product: I-O DATA GV-USB2
[  226.724106] usb 1-3: Manufacturer: I-O DATA
[  226.724107] usb 1-3: SerialNumber: 000000000000001
...
[  226.772110] usbcore: registered new interface driver gvusb2-snd
[  226.776674] mc: Linux media interface: v0.10
[  226.787110] videodev: Linux video capture interface: v2.00
[  226.792728] gvusb2-vid 1-3:1.0: found video at altsetting 5 endpoint 1
[  227.071151] usbcore: registered new interface driver gvusb2-vid
```

I tested with kernel `5.10.7`:
```
    boot.extraModulePackages = with linuxPackages_latest; [
      gvusb2
    ];
```

`vfl_type_grabber.patch` addresses https://github.com/Isaac-Lozano/GV-USB2-Driver/issues/2 on appropriate kernel versions.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by djanatyn** [comment]
I saw that the upstream repository had two licenses in the tree:
* https://github.com/Isaac-Lozano/GV-USB2-Driver/blob/811fb0f6ee4ef62e0e346a293bb5e3b61ad10b7d/LICENSE.bsd
* https://github.com/Isaac-Lozano/GV-USB2-Driver/blob/811fb0f6ee4ef62e0e346a293bb5e3b61ad10b7d/LICENSE.gpl

I have the license set to GPL. Maybe it should be set to both? I'm not sure what should apply.

**Comment by SuperSandro2000** [comment]
> I saw that the upstream repository had two licenses in the tree:
> 
> * [Isaac-Lozano/GV-USB2-Driver@`811fb0f`/LICENSE.bsd](https://github.com/Isaac-Lozano/GV-USB2-Driver/blob/811fb0f6ee4ef62e0e346a293bb5e3b61ad10b7d/LICENSE.bsd)
> * [Isaac-Lozano/GV-USB2-Driver@`811fb0f`/LICENSE.gpl](https://github.com/Isaac-Lozano/GV-USB2-Driver/blob/811fb0f6ee4ef62e0e346a293bb5e3b61ad10b7d/LICENSE.gpl)
> 
> I have the license set to GPL. Maybe it should be set to both? I'm not sure what should apply.

They apply both.

File: pkgs/os-specific/linux/gvusb2/Makefile.patch [hunk 1]
```diff
@@ -0,0 +1,17 @@
+diff --git a/Makefile b/Makefile
+index e666392..e1a341b 100644
+--- a/Makefile
++++ b/Makefile
+@@ -3,8 +3,10 @@ obj-m += gvusb2-sound.o gvusb2-video.o
+ gvusb2-sound-y := gvusb2-snd.o gvusb2-core.o
+ gvusb2-video-y := gvusb2-vid.o gvusb2-v4l2.o gvusb2-core.o gvusb2-i2c.o
+ 
++KERNEL_SOURCE_DIR := /lib/modules/$(shell uname -r)/build
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/gvusb2/Makefile.patch:9]: Why do we need this patch? It looks like it changes nothing for me.

File: pkgs/os-specific/linux/gvusb2/default.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  installPhase = ''
+    install -D {,$out/${kerneldir}/extra/}gvusb2-sound.ko
+    install -D {,$out/${kerneldir}/extra/}gvusb2-video.ko
+  '';
+
+  meta = with lib; {
+    description = "A linux driver for the IO-DATA GV-USB2 SD capture device";
+    homepage = "https://github.com/Isaac-Lozano/GV-USB2-Driver";
+    license = licenses.gpl2;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/gvusb2/default.nix:35]:
```suggestion
    license = with licenses [ bsd3 gpl2Only ];
```

File: pkgs/os-specific/linux/gvusb2/default.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    "INSTALL_MOD_PATH=$(out)"
+  ];
+
+  patches = with lib; [
+    ./Makefile.patch
+  ] ++ (optional (versionOlder kernel.version "5.7") ./vfl_type_grabber.patch);
+
+  installPhase = ''
+    install -D {,$out/${kerneldir}/extra/}gvusb2-sound.ko
+    install -D {,$out/${kerneldir}/extra/}gvusb2-video.ko
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/gvusb2/default.nix:29]:
```suggestion
    install -D gvusb2-sound.ko $out/${kerneldir}/extra/gvusb2-sound.ko
    install -D gvusb2-video.ko $out/${kerneldir}/extra/gvusb2-video.ko
```

File: pkgs/os-specific/linux/gvusb2/default.nix [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  nativeBuildInputs = kernel.moduleBuildDependencies;
+
+  makeFlags = [
+    "KERNEL_SOURCE_DIR=${kernel.dev}/${kerneldir}/build"
+    "INSTALL_MOD_PATH=$(out)"
+  ];
+
+  patches = with lib; [
+    ./Makefile.patch
+  ] ++ (optional (versionOlder kernel.version "5.7") ./vfl_type_grabber.patch);
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/gvusb2/default.nix:25]:
```suggestion
  patches = [
    ./Makefile.patch
  ] ++ (lib.optional (lib.versionOlder lib.kernel.version "5.7") ./vfl_type_grabber.patch);
```

**Comment by SuperSandro2000** [comment]
```
  trying https://github.com/Isaac-Lozano/gvusb2/archive/811fb0f6ee4ef62e0e346a293bb5e3b61ad10b7d.tar.gz

% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
Dload  Upload   Total   Spent    Left  Speed
0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
error: cannot download source from any mirror
```

**Comment by djanatyn** [comment]
> ```
>   trying https://github.com/Isaac-Lozano/gvusb2/archive/811fb0f6ee4ef62e0e346a293bb5e3b61ad10b7d.tar.gz
> 
> % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
> Dload  Upload   Total   Spent    Left  Speed
> 0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> curl: (22) The requested URL returned error: 404
> error: cannot download source from any mirror
> ```

Sorry, I changed `pname` without rebuilding. I pushed the correct repo name: https://github.com/NixOS/nixpkgs/pull/109560/commits/657b7b899f5bcd8bd5fa3416219826706337cbd4

File: pkgs/os-specific/linux/gvusb2/Makefile.patch (hunk 1 above)
Comment (djanatyn) [reply pkgs/os-specific/linux/gvusb2/Makefile.patch:9]: The Makefile included with the project doesn't use any environment variables:
```
❯ git show 811fb0f6ee4ef62e0e346a293bb5e3b61ad10b7d:Makefile
obj-m += gvusb2-sound.o gvusb2-video.o

gvusb2-sound-y := gvusb2-snd.o gvusb2-core.o
gvusb2-video-y := gvusb2-vid.o gvusb2-v4l2.o gvusb2-core.o gvusb2-i2c.o

all:
        make -C /lib/modules/$(shell uname -r)/build M=$(PWD) modules

clean:
        make -C /lib/modules/$(shell uname -r)/build M=$(PWD) clean

```

Without this patch, the build is running the default `make -C /lib/modules/5.10.7/build M=/build/source modules` and failing:
```
❯ nix-build -A linuxPackages_latest.gvusb2
these derivations will be built:
  /nix/store/warcxpjpb5kknzkwrnbq21b50wavbz1k-gvusb2-2020-07-27-unstable.drv
building '/nix/store/warcxpjpb5kknzkwrnbq21b50wavbz1k-gvusb2-2020-07-27-unstable.drv'...
unpacking sources
unpacking source archive /nix/store/iwc4l1vs3sqxmf99cq0839dl2zh2xnn1-source
source root is source
patching sources
configuring
no configure script, doing nothing
building
build flags: SHELL=/nix/store/kl6lr3czkbnr6m5crcy8ffwfzbj8a22i-bash-4.4-p23/bin/bash KERNEL_SOURCE_DIR=/nix/store/2kxfgzzdwi1nj6axsi43r05mcd6ncqg4-linux-5.10.7-dev/lib/modules/5.10.7/build INSTALL_MOD_PATH=\$\(out\)
make -C /lib/modules/5.10.7/build M=/build/source modules
make[1]: *** /lib/modules/5.10.7/build: No such file or directory.  Stop.
make: *** [Makefile:7: all] Error 2
builder for '/nix/store/warcxpjpb5kknzkwrnbq21b50wavbz1k-gvusb2-2020-07-27-unstable.drv' failed with exit code 2
error: build of '/nix/store/warcxpjpb5kknzkwrnbq21b50wavbz1k-gvusb2-2020-07-27-unstable.drv' failed
```

I'm using https://github.com/NixOS/nixpkgs/pull/105899 as a reference along with https://nixos.wiki/wiki/Linux_kernel#Packaging_out-of-tree_kernel_modules.

I'm adding an environment variable `KERNEL_SOURCE_DIR` to the Makefile to choose the right directory during the build, using `${kernel.dev}`:
`make -C /nix/store/2kxfgzzdwi1nj6axsi43r05mcd6ncqg4-linux-5.10.7-dev/lib/modules/5.10.7/build M=/build/source modules`
```
build flags: SHELL=/nix/store/kl6lr3czkbnr6m5crcy8ffwfzbj8a22i-bash-4.4-p23/bin/bash KERNEL_SOURCE_DIR=/nix/store/2kxfgzzdwi1nj6axsi43r05mcd6ncqg4-linux-5.10.7-dev/lib/modules/5.10.7/build INSTALL_MOD_PATH=\$\(out\)
make -C /nix/store/2kxfgzzdwi1nj6axsi43r05mcd6ncqg4-linux-5.10.7-dev/lib/modules/5.10.7/build M=/build/source modules
```

I don't have much experience building kernel modules and may be missing some conventions here.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 109560` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages marked as broken and skipped:
</summary>
<ul>
<li>linuxPackages_hardkernel_4_14.gvusb2</li>
<li>linuxPackages_hardkernel_latest.gvusb2</li>
</ul>
</details>
<details>

  <summary>2 packages failed to build and are new build failures:</summary>
  <ul>
<li>linuxPackages-libre.gvusb2: log was empty</li>
<li>linuxPackages_4_4.gvusb2: log <a href="https://termbin.com/d9iv">https://termbin.com/d9iv</a></li>
  </ul>
</details>
<details>
<summary>
16 packages built:
</summary>
<ul>
<li>linuxPackages.gvusb2 (linuxPackages_5_4.gvusb2)</li>
<li>linuxPackages_4_14.gvusb2</li>
<li>linuxPackages_4_19.gvusb2</li>
<li>linuxPackages_4_9.gvusb2</li>
<li>linuxPackages_5_10.gvusb2 (linuxPackages_latest.gvusb2)</li>
<li>linuxPackages_5_9.gvusb2</li>
<li>linuxPackages_hardened.gvusb2</li>
<li>linuxPackages_latest-libre.gvusb2</li>
<li>linuxPackages_latest_hardened.gvusb2</li>
<li>linuxPackages_latest_xen_dom0.gvusb2</li>
<li>linuxPackages_latest_xen_dom0_hardened.gvusb2</li>
<li>linuxPackages_lqx.gvusb2</li>
<li>linuxPackages_testing_bcachefs.gvusb2</li>
<li>linuxPackages_xen_dom0.gvusb2</li>
<li>linuxPackages_xen_dom0_hardened.gvusb2</li>
<li>linuxPackages_zen.gvusb2</li>
</ul>
</details>

**Comment by Artturin** [comment]
Reopen and rebase if you're still interested in this.

**Pull request closed without merging**